### DISCUSSION TITLE
ClientException when sending request to custom endpoint

### DISCUSSION THREAD

**Discussion by aluedtke7** [pocketbase/pocketbase#1833] [Q&A] [answered]
Hi,
I've extended pocketbase with a custom endpoint (simple hello world example from the docs):
```
	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodGet,
			Path:   "/hello",
			Handler: func(c echo.Context) error {
				log.Print("/hello was called")
				return c.String(http.StatusOK, "Hello world!")
			},
			Middlewares: []echo.MiddlewareFunc{
				apis.ActivityLogger(app),
				apis.RequireRecordAuth(),
			},
		})
		return nil
	})
```
I'm also using the Dart Sdk in a Flutter app to access pocketbase. There, I call the endpoint like this:
```
PocketBase pb = PocketBase('http://127.0.0.1:8090');
await pb.collection('users').authWithPassword(email, password);
pb.send('/hello').then((value) {
      debugPrint('then $value');
}).onError((error, stackTrace) {
      debugPrint('error $error');
});
```
Doing so, gives me on the client (Flutter app) this exception:
```
flutter: error  ClientException: {url: http://127.0.0.1:8090/hello, isAbort: false, statusCode: 0, response: {}, originalError: FormatException: Unexpected character (at character 1)
flutter: Hello world!
flutter: ^
flutter: }
```
Why do I get a **FormatException**? Which Format is the client expecting?
When I send the same request from Postman, I get the string "Hello world!" without any error (status code is 200).

**Comment by ganigeorgiev** [comment, marked as answer]
This is because the SDK `send()` method expects the response body to be either empty (eg. in case of 204) or a json.

To return a json from your custom endpoint, you can call `c.JSON`, eg:

```go
c.JSON(http.StatusOK, map[string]any{"test": 123})`
```

_I'll consider documenting it better with the next SDK release (or maybe silently ignore json decode error and return the response as it is)._

**Reply by aluedtke7** [reply]
Thanks, for your fast response!

**Reply by ganigeorgiev** [reply]
_Just to let you know, I've pushed a [fix in v0.7.2](https://github.com/pocketbase/dart-sdk/commit/7a283531be4355862abd0c6fcf8e3b987f737745) that should allow processing non-json responses. Depending on the users demand, we may also introduce in the future other data normalizations based on the response body `Content-Type` header, but for now in case the `jsonDecode` fails it will return the response as it is (for consistency the same behavior eventually will be applied also to the JS SDK with its next release)._

**Reply by aluedtke7** [reply]
You are truly amazing! 

I reported my problem here and 35 minutes later you responded with the answer. Another 45 minutes later, you patched the sdk and created a new release. 
And all this happened on a saturday afternoon!
Thank you very much
